### ISSUE TITLE
[apollo-datasource-rest] Possible memory leak due to TTL failure

### ISSUE THREAD

**Issue by Huljo** [apollographql/datasource-rest#44]
# Description

We're using apollo-datasource-rest alongside apollo-server to manage external REST requests.

We noticed that using any TTL value when doing a request leads to several memory leaks.

Here is an allocation timeline of the server while using `{cacheOptions: {ttl: 1}}`
![cx-live-snapshots-1s-memory-cache](https://user-images.githubusercontent.com/12231164/81499022-a9e9ae80-92c8-11ea-921a-faa580ebba96.gif)

Timeline using `{cacheOptions: {ttl: -1}}`
![cx-live-snapshots-no-memory-cache](https://user-images.githubusercontent.com/12231164/81499051-e2898800-92c8-11ea-9a1e-ec8371a7ba6d.gif)

We used `1` for this example but any other value creates same leaks in our observations.

The memory usage is quite stable in case of TTL `-1`:
![cx-no-memory-cache-monitoring](https://user-images.githubusercontent.com/12231164/81499115-6e031900-92c9-11ea-9625-bad01c584e81.png)

But still growing in case of TTL `1`: 
![cx-1s-memory-cache-monitoring](https://user-images.githubusercontent.com/12231164/81499125-7eb38f00-92c9-11ea-970c-62362829543f.png)

 ### Version Info
 
```
apollo-server: 2.13.0
apollo-datasource-rest: 0.9.0
```

## Expected Behavior

When TTL = `1` I'm expecting cached requests to be removed from memory after 1 sec but the InMemoryLRUCache store is still growing in my case.
![memory-store-detail](https://user-images.githubusercontent.com/12231164/81500084-fe445c80-92cf-11ea-9d52-d236dce56ba9.png)

### Steps

```shell
yarn i
yarn run build
node --inspect dist/index.js
# open chrome://inspect in chrome
# send queries
```

Here is the server we use: 
````javascript
# index.js
const server = new ApolloServer({
    schema: MySchema,
    dataSources: {
       myAPI: new MyAPI({baseURL: ...})
    },
    context: ...,
    cacheControl: {
      defaultMaxAge: 86400,
    },
  });

````

Datasource: 
````javascript
export default class MyAPI extends RESTDataSource {
  constructor(options) {
    super();
    this.baseURL = options.baseURL;
  }

  async getItem(id) {
    const apiResult = await this.get(
      `/MyEndpointPath/${id}`,
      {
        aGetParam: 'getParamValue',
      },
      { cacheOptions: { ttl: 1 } },
    );
  }
  return {
     name: apiResult.title
  };
}

````

Resolver
````javascript
export default {
  Query: {
       getItem(_, args, context): object {
              return context.dataSources.myAPI.getItem(args.id);
         },
   }
}
````

Schema
````
extend type Query {
  getItem(id: String): Item
}

type Item @cacheControl(maxAge: 86400) {
    name: String
}
````

Is it expected behavior or did we miss a part of the configuration ?

**Comment by glasser** [comment]
I'd love to investigate this now that we are maintaining RESTDataSource again, but the original report isn't a complete reproduction recipe that allows us to try it out without a bunch of creative choices to try to reproduce what you were seeing. The code snippets are incomplete, and also it doesn't say what kind of operations you run (the same operation over and over? different ones?)

To be honest, I would expect exactly the opposite issue here: back in 2020 when you filed this, the default KeyValueCache used by ApolloServer had an unbounded size, so without any TTL I *would* expect a leak, where as with a TTL I would expect no leak.  Looking into it further, it looks like `lru-cache@6` (which powered InMemoryLRUCache back then) doesn't have any special handling of a `ttl` of -1 (which we convert to a maxAge of -1000) and this is interpreted not as "there is no TTL, keep forever" but "it's already out of date, drop as soon as you bother to check". So it's not too surprising that `ttl: 1` uses more memory than `ttl: -1`.

Additionally, the `lru-cache` library does not actually active purge things that are out of date by default. It's just that if you try to fetch something that is out of date, it will get dropped. The older lru-cache@6 version used when this was filed had no way of automatically dropping old data other than manually calling `prune()` every so often. The newer version used by today's InMemoryLRUCache does have a `ttlAutopurge` flag, though it's not recommended by the author. More importantly, the newer version of `lru-cache` doesn't let you configure an unbounded cache, and so the current version of InMemoryLRUCache is always bounded. 

So basically, I think everything you're seeing here is addressed by the fact that the current InMemoryLRUCache (in `@apollo/utils.keyvaluecache`) is backed by lru-cache v7, which is always a bounded cache.

If you're able to reproduce this with the latest versions of RESTDataSource (in `@apollo/datasource-rest`, not `apollo-dataousrce-rest`) we can reopen (ideally with a full self-contained reproduction!).
